torchvision/models/detection/ssd.py

```diff
@@ -0,0 +1,486 @@
+        if isinstance(layer, nn.Conv2d):
+            torch.nn.init.xavier_uniform_(layer.weight)
+            if layer.bias is not None:
+                torch.nn.init.constant_(layer.bias, 0.0)
+
+
+class SSDHead(nn.Module):
+    def __init__(self, in_channels: List[int], num_anchors: List[int], num_classes: int):
+        super().__init__()
+        self.classification_head = SSDClassificationHead(in_channels, num_anchors, num_classes)
+        self.regression_head = SSDRegressionHead(in_channels, num_anchors)
+
+    def forward(self, x: List[Tensor]) -> Dict[str, Tensor]:
+        return {
+            'bbox_regression': self.regression_head(x),
+            'cls_logits': self.classification_head(x),
+        }
+
+
+class SSDScoringHead(nn.Module):
+    def __init__(self, module_list: nn.ModuleList, num_columns: int):
+        super().__init__()
+        self.module_list = module_list
+        self.num_columns = num_columns
+
+    def _get_result_from_module_list(self, x: Tensor, idx: int) -> Tensor:
+        """
+        This is equivalent to self.module_list[idx](x),
+        but torchscript doesn't support this yet
+        """
+        num_blocks = len(self.module_list)
+        if idx < 0:
+            idx += num_blocks
+        i = 0
+        out = x
+        for module in self.module_list:
+            if i == idx:
+                out = module(x)
+            i += 1
+        return out
+
+    def forward(self, x: List[Tensor]) -> Tensor:
+        all_results = []
+
+        for i, features in enumerate(x):
+            results = self._get_result_from_module_list(features, i)
+
+            # Permute output from (N, A * K, H, W) to (N, HWA, K).
+            N, _, H, W = results.shape
+            results = results.view(N, -1, self.num_columns, H, W)
+            results = results.permute(0, 3, 4, 1, 2)
+            results = results.reshape(N, -1, self.num_columns)  # Size=(N, HWA, K)
+
+            all_results.append(results)
+
+        return torch.cat(all_results, dim=1)
+
+
+class SSDClassificationHead(SSDScoringHead):
+    def __init__(self, in_channels: List[int], num_anchors: List[int], num_classes: int):
+        cls_logits = nn.ModuleList()
+        for channels, anchors in zip(in_channels, num_anchors):
+            cls_logits.append(nn.Conv2d(channels, num_classes * anchors, kernel_size=3, padding=1))
+        _xavier_init(cls_logits)
+        super().__init__(cls_logits, num_classes)
+
+
+class SSDRegressionHead(SSDScoringHead):
+    def __init__(self, in_channels: List[int], num_anchors: List[int]):
+        bbox_reg = nn.ModuleList()
+        for channels, anchors in zip(in_channels, num_anchors):
+            bbox_reg.append(nn.Conv2d(channels, 4 * anchors, kernel_size=3, padding=1))
+        _xavier_init(bbox_reg)
+        super().__init__(bbox_reg, 4)
+
+
+class SSD(nn.Module):
+    __annotations__ = {
+        'box_coder': det_utils.BoxCoder,
+        'proposal_matcher': det_utils.Matcher,
```

Comment:
Strangely JIT complains if I declare `DBoxMatcher` instead here. Works fine if I use the base class.

Reply:
Would be worth opening an issue in PyTorch about this.

Additionally, I'm not even sure we need the `__annotations__` anymore with latest torchscript, might be good to double-check in the future. At the very least, we can use Python3 type annotations in the constructor

Reply:
Removing the `__annotations__` leads to the following exception while running the test. To make it easier to describe/reproduce the issue, I think I'll wait for this to merge and then open an issue with PyTorch asking them whether I should be receiving an exception after removing the annotations. 

```
FAILED      [ 33%]
test/test_models.py:446 (test_detection_model[dev0-ssd300_vgg16])
model_name = 'ssd300_vgg16', dev = device(type='cpu')

    @pytest.mark.parametrize('model_name', get_available_detection_models())
    @pytest.mark.parametrize('dev', _devs)
    def test_detection_model(model_name, dev):
>       ModelTester()._test_detection_model(model_name, dev)

test/test_models.py:450: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_models.py:217: in _test_detection_model
    self.check_jit_scriptable(model, ([x],), unwrapper=script_model_unwrapper.get(name, None))
test/common_utils.py:273: in check_jit_scriptable
    sm = torch.jit.script(nn_module)
python3.8/site-packages/torch/jit/_script.py:911: in script
    return torch.jit._recursive.create_script_module(
python3.8/site-packages/torch/jit/_recursive.py:370: in create_script_module
    return create_script_module_impl(nn_module, concrete_type, stubs_fn)
python3.8/site-packages/torch/jit/_recursive.py:430: in create_script_module_impl
    create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

concrete_type = <torch._C.ConcreteModuleType object at 0x7fb60937bfb0>
method_stubs = [ScriptMethodStub(resolution_callback=<function createResolutionCallbackFromEnv.<locals>.<lambda> at 0x7fb5f8deaca0>, ...(mean=[123.0, 117.0, 104.0], std=[1.0, 1.0, 1.0])
      Resize(min_size=(300,), max_size=300, mode='bilinear')
  )
)>)]
property_stubs = []

    def create_methods_and_properties_from_stubs(concrete_type, method_stubs, property_stubs):
        method_defs = [m.def_ for m in method_stubs]
        method_rcbs = [m.resolution_callback for m in method_stubs]
        method_defaults = [get_default_args(m.original_method) for m in method_stubs]
    
        property_defs = [p.def_ for p in property_stubs]
        property_rcbs = [p.resolution_callback for p in property_stubs]
    
>       concrete_type._create_methods_and_properties(property_defs, property_rcbs, method_defs, method_rcbs, method_defaults)
E       RuntimeError: 
E       Module 'SSD' has no attribute 'proposal_matcher' (This attribute exists on the Python module, but we failed to convert Python type: 'torchvision.models.detection._utils.DBoxMatcher' to a TorchScript type.):
E         File "./vision/torchvision/models/detection/ssd.py", line 290
E           
E                       match_quality_matrix = box_ops.box_iou(targets_per_image['boxes'], anchors_per_image)
E                       matched_idxs.append(self.proposal_matcher(match_quality_matrix))
E                                           ~~~~~~~~~~~~~~~~~~~~~ <--- HERE
E           
E                   losses = self.compute_loss(targets, head_outputs, anchors, matched_idxs)

python3.8/site-packages/torch/jit/_recursive.py:322: RuntimeError
```